Orange/widgets/utils/owlearnerwidget.py

```diff
@@ -1,26 +1,229 @@
+import numpy as np
+
+from Orange.classification.base_classification import LearnerClassification
+from Orange.data import Table
 from Orange.preprocess.preprocess import Preprocess
+from Orange.regression.base_regression import LearnerRegression
+from Orange.util import abstract
+from Orange.widgets import gui
+from Orange.widgets.settings import Setting
+from Orange.widgets.utils.sql import check_sql_input
+from Orange.widgets.widget import OWWidget, WidgetMetaClass
+
+
+class DefaultWidgetChannelsMetaClass(WidgetMetaClass):
+    """Metaclass that adds default inputs and outputs objects.
+    """
+
+    REQUIRED_ATTRIBUTES = []
+
+    def __new__(mcls, name, bases, attrib):
+        # check whether it is abstract class
+        if attrib.get('name', False):
+            # Ensure all needed attributes are present
+            if not all(attr in attrib for attr in mcls.REQUIRED_ATTRIBUTES):
+                raise AttributeError("'{name}' must have '{attrs}' attributes"
+                                     .format(name=name, attrs="', '".join(mcls.REQUIRED_ATTRIBUTES)))
+
+            attrib['outputs'] = mcls.update_channel(
+                mcls.default_outputs(attrib),
+                attrib.get('outputs', [])
+            )
+
+            attrib['inputs'] = mcls.update_channel(
+                mcls.default_inputs(attrib),
+                attrib.get('inputs', [])
+            )
+
+            mcls.add_extra_attributes(name, attrib)
+
+        return super().__new__(mcls, name, bases, attrib)
+
+    @classmethod
+    def default_inputs(cls, attrib):
+        return []
+
+    @classmethod
+    def default_outputs(cls, attrib):
+        return []
 
+    @classmethod
+    def update_channel(cls, channel, items):
+        item_names = set(item[0] for item in channel)
 
-class _OWAcceptsPreprocessor:
+        for item in items:
+            if not item[0] in item_names:
+                channel.append(item)
+
+        return channel
+
+    @classmethod
+    def add_extra_attributes(cls, name, attrib):
+        return attrib
+
+
+class DefaultLearnerWidgetChannelsMetaClass(DefaultWidgetChannelsMetaClass):
```

Comment:
Can you just rename this to something more up-to-date, like `OWBaseLearnerMeta`.